pdr_backend/sim/sim_engine.py

```diff
@@ -261,3 +291,113 @@ def save_state(self, i: int, N: int):
+        query_cont = f"""
+            SELECT
+                contract
+            FROM
+                pdr_predictions
+            WHERE
+                timeframe = '{self.predict_feed.timeframe}'
+                AND pair = '{self.predict_feed.pair.pair_str}'
+            LIMIT 1
+        """
+        feed_contract_addr = db.query_data(query_cont)
+        if len(feed_contract_addr) == 0:
+            logger.error("Contract address for the given feed not found in database")
+            return result_dict
+
+        query = f"""
+            SELECT
+                slot,
+                CASE
+                    WHEN roundSumStakes = 0.0 THEN NULL
+                    ELSE roundSumStakesUp / roundSumStakes
+                END AS probUp
+            FROM
+                pdr_payouts
+            WHERE
+                slot > {start_slot}
+                AND slot < {end_slot}
+                AND ID LIKE '{feed_contract_addr.row(0)[0]}%'
+        """
+
+        df: pl.DataFrame = db.query_data(query)
+
+        for i in range(len(df)):
+            if df["probUp"][i] is not None:
+                result_dict[df["slot"][i]] = df["probUp"][i]
+
+        return result_dict
+
+    def _get_past_predictions_from_chain(self, ppss: PPSS):
+        # calculate needed data start date
+        current_time_s = int(time.time())
+        timeframe = ppss.trader_ss.feed.timeframe
+        number_of_data_points = ppss.sim_ss.test_n
+        start_date = current_time_s - (timeframe.s * number_of_data_points)
+
+        # check if ppss is correctly configured for data ferching
+        if start_date < int(
+            UnixTimeMs.from_timestr(self.ppss.lake_ss.st_timestr) / 1000
+        ):
+            logger.info(
+                (
+                    "Lake dates configuration doesn't meet the requirements. "
+                    "Make sure you set start date before %s"
+                ),
+                time.strftime("%Y-%m-%d", time.localtime(start_date)),
+            )
+            return False
+
+        # fetch data from subgraph
+        try:
+            gql_data_factory = GQLDataFactory(ppss)
+            gql_data_factory._update()
+        except Exception as e:
+            logger.error("Fetching chain data failed. %s", e)
+
+        # check if required data exists in the data base
+        db = DuckDBDataStore(self.ppss.lake_ss.lake_dir)
+        query = """
+        (SELECT timestamp
+            FROM pdr_payouts
+            ORDER BY timestamp ASC
+            LIMIT 1)
+            UNION ALL
+            (SELECT timestamp
+            FROM pdr_payouts
+            ORDER BY timestamp DESC
+            LIMIT 1);
+        """
+        data = db.query_data(query)
+
+        if (data is not None) and (len(data["timestamp"]) < 2):
+            logger.info(
+                "No prediction data found in database at %s", self.ppss.lake_ss.lake_dir
+            )
+            return False
+        start_timestamp = data["timestamp"][0] / 1000
```

Comment:
No multiplication or division by 1000 outside of time utils! Use the conversions there or place them there.